manager/controllers/app/blueprint_controller.go

```diff
@@ -51,41 +52,88 @@ func (r *BlueprintReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+
 	result, err := r.reconcile(ctx, log, &blueprint)
 	if err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "failed to reconcile blueprint")
 	}
 
-	if blueprint.DeletionTimestamp == nil {
-		if !equality.Semantic.DeepEqual(&blueprint.Status, observedStatus) {
-			if err := r.Client.Status().Update(ctx, &blueprint); err != nil {
-				return ctrl.Result{}, errors.WrapWithDetails(err, "failed to update blueprint status", "status", blueprint.Status)
-			}
+	if !equality.Semantic.DeepEqual(&blueprint.Status, observedStatus) {
+		if err := r.Client.Status().Update(ctx, &blueprint); err != nil {
+			return ctrl.Result{}, errors.WrapWithDetails(err, "failed to update blueprint status", "status", blueprint.Status)
 		}
 	}
 
 	log.Info("blueprint reconcile cycle completed", "result", result)
 	return result, nil
 }
 
-func (r *BlueprintReconciler) deleteChartResource(ctx context.Context, kubeNamespace string, releaseName string) (ctrl.Result, error) {
-	rel, err := r.Helmer.Status(kubeNamespace, releaseName)
-	if err == nil && rel != nil {
-		_, _ = r.Helmer.Uninstall(kubeNamespace, releaseName)
+// reconcileFinalizers reconciles finalizers for Blueprint
+func (r *BlueprintReconciler) reconcileFinalizers(blueprint *app.Blueprint) error {
+	// finalizer
+	finalizerName := r.Name + ".finalizer"
+	hasFinalizer := ctrlutil.ContainsFinalizer(blueprint, finalizerName)
+
+	// If the object has a scheduled deletion time, delete it and its associated resources
+	if !blueprint.DeletionTimestamp.IsZero() {
+		// The object is being deleted
+		if hasFinalizer { // Finalizer was created when the object was created
+			// the finalizer is present - delete the allocated resources
+			if err := r.deleteExternalResources(blueprint); err != nil {
+				return err
+			}
+
+			// remove the finalizer from the list and update it, because it needs to be deleted together with the object
+			ctrlutil.RemoveFinalizer(blueprint, finalizerName)
+
+			if err := r.Update(context.Background(), blueprint); err != nil {
+				return err
+			}
+		}
+		return nil
 	}
+	// Make sure this CRD instance has a finalizer
+	if !hasFinalizer {
+		ctrlutil.AddFinalizer(blueprint, finalizerName)
+		if err := r.Update(context.Background(), blueprint); err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-	return ctrl.Result{}, nil
+func getReleaseName(step app.FlowStep) string {
+	// we add the "r" character at the beginning of the release name, since it must begin with an alphabetic character
+	return "r" + utils.Hash(step.Name, 20)
+}
+
+func (r *BlueprintReconciler) deleteExternalResources(blueprint *app.Blueprint) error {
+	var err error
+	for _, step := range blueprint.Spec.Flow.Steps {
+		releaseName := getReleaseName(step)
+		if rel, errStatus := r.Helmer.Status(blueprint.Namespace, releaseName); errStatus != nil || rel == nil {
+			continue
+		}
+		_, err = r.Helmer.Uninstall(blueprint.Namespace, releaseName)
+	}
+	return err
```

Comment:
1. Should we log errStatus for debugging?
2. This implementation simply returns what the last call to uninstall returns, which might be nil even if a previous uninstall call returned an error. You can either fix this directly (e.g., to only assign to the err var on the first actual error), or consider to group all returned errors (see implementation in https://github.com/helm/helm/blob/master/pkg/action/uninstall.go or the https://github.com/hashicorp/go-multierror library).

Reply:
Fixed.